### DISCUSSION TITLE
Using open telemetry in the compiler

### DISCUSSION THREAD

**Discussion by nojaf** [dotnet/fsharp#15340] [Q&A]
Hi all, 

Earlier today I was debugging `IncrementalBuild.fs` and was feeling quite overwhelmed and confused about where the code was flowing. Then it hit, I could try and view some of the open telemetry traces that were added in https://github.com/dotnet/fsharp/pull/13835.

I really a noob when it comes to otel. I understand what it is, but have never really played with it myself. Never did Jaegher, Zipkin or other stuff, so I was pretty clueless about how to enable it during my unit test.

After reading some threads I found out that it is quite easy to connect to [otel-desktop-viewer](https://github.com/CtrlSpice/otel-desktop-viewer).

I downloaded the latest (Windows) [release](https://github.com/CtrlSpice/otel-desktop-viewer/releases/tag/v0.1.2) and started the executable.

Next, I added [OpenTelemetry.Exporter.OpenTelemetryProtocol](https://www.nuget.org/packages/OpenTelemetry.Exporter.OpenTelemetryProtocol) to  `FSharp.Compiler.ComponentTests.fsproj` and hook it up using:

```fsharp
open System
open OpenTelemetry
open OpenTelemetry.Trace
open OpenTelemetry.Exporter
open OpenTelemetry.Resources

[<Fact>]
let ``Some test `` () =
    use tracerProvider =
        Sdk.CreateTracerProviderBuilder()
             // FSharp.Compiler.Diagnostics.ActivityNames.AllRelevantNames
             .AddSource([| "fsc"; "fsc_with_env_stats" |])
             .SetResourceBuilder(
                ResourceBuilder
                    .CreateDefault()
                    .AddService (serviceName = "Debug FCS")
            )
            .AddOtlpExporter(fun (options:OtlpExporterOptions) ->
                // otel-desktop-viewer is spinning on that port.
                options.Endpoint <- Uri("http://localhost:4317/")
                options.Protocol <- OtlpExportProtocol.Grpc
            )
            .Build();
        
    Fs "module Foo"
    |> compile
    |> shouldSucceed
```

and I started seeing traces like:

![traces](https://github.com/dotnet/fsharp/assets/2621499/dbc3e52b-5558-475d-8c73-296673bcf6e6)

This is pretty useful and so my open questions about this are:
- Can we have `OpenTelemetry.Exporter.OpenTelemetryProtocol` in test projects? And maybe some helpers to easily ad-hoc use this.
- Could we have something that also would do this for `fsc.fsproj`? Maybe a second project somewhere that users can tweak to have otel during local compilations.
- Could we have some documentation on the otel things in the compiler?

**Comment by baronfel** [comment]
I love that viewer, it's the simplest way to get up and running with otel <3 It's what I was hoping to use with a toggle in Ionide to transparently start collecting traces.

For your code sample, you shouldn't need to provide any parameters to the otel exporter because those are the defaults, so that should make the hooking up even simpler.

**Comment by 0101** [comment]
Did you see this PR? https://github.com/dotnet/fsharp/pull/15223

Jaeger is similarly easy to use, just download & run. I didn't try any other viewers because I was happy with it. And it's already included in test projects. But we could include the other one as well, or if it's better just use that one. 

Don't know about putting it in `fcs.fsproj`, but maybe a second project would be a solution there. 

I would also really like to hook it up to VS locally but didn't figure that out yet.

And yes, we should add documentation for it :)

**Reply by nojaf** [reply]
I did not. Shame this got merged without any actual documentation. 
Does the `Jaeger` thing just work, or was there any setup in the `ProjectWorkflowBuilder` done?

**Reply by 0101** [reply]
Yeah, I know, I should add at least something in terms of docs... 

You can see the setup in that PR. It's essentially the same you have. Just ensures that we start collecting only after the initial setup and checking of the project, so you only see the events of your actual test.

**Comment by vzarytovskii** [comment]
Regarding fsc project - we shouldn't have anything there, as it's a part of shipping product.

Separate project or dynamic loading of exporter might work as well.

**Comment by cartermp** [comment]
Hey, otel maintainer here -- I would say that an exporter in the fsc project makes sense and would be great to have. The overhead is likely marginal and it'd be great to just be able to run the compiler and whatever local tool you have picks it up so you can see a full compile run.

**Reply by vzarytovskii** [reply]
Problem is that VS likely won't allow any 3rd party DLLs, or getting an approval will be painful.

**Reply by vzarytovskii** [reply]
Ability to load a DLL for exporter on startup sounds feasible though.

**Reply by cartermp** [reply]
well otel-dotnet is almost exclusively built by microsoft, so much so that they basically own the whole project :)
